File: javascript/leetcode/814. Binary Tree Pruning.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var pruneTree = function(root) {
  if (!root) return null;

  root.right = pruneTree(root.right);
  root.left = pruneTree(root.left);

  if (root.val !== 1 && !root.left && !root.right) return null;

  return root;
};

File: c/leetcode/814. Binary Tree Pruning.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *pruneTree(struct TreeNode *root)
{
  if (!root)
    return NULL;

  root->right = pruneTree(root->right);
  root->left = pruneTree(root->left);

  if (root->val != 1 && !root->right && !root->left)
    return NULL;

  return root;
}
